keps/sig-storage/1710-selinux-relabeling/README.md

```diff
@@ -818,26 +1054,166 @@ With a single `VolumeChangePolicy`, user has to fall back to `Recursive` policy
+The Pod B will be `ContainerCreating` until Pod A terminates and the volume can be mounted with the right SELinux label.
 The same behavior was already implemented for `ReadWriteOncePod` AccessMode.
+
+### Mount **all** volumes with `-o context`, without any `SELinuxChangePolicy`
+
+Initially, we wanted kubelet to do the right thing automatically, without any user intervention.
+Later found out that sharing the same volume among privileged and unprivileged containers is a valid use case.
+That use case would not be possible if all mount used `-o context`:
+
+1. An privileged Pod A runs. Its volumes are mounted without any special `-o context`, privileged pods run with `spc_t` label.
+2. An unprivileged Pod B with label `c1,c2` is about to start, and it wants to use volume already mounted to Pod A.
+   Kubelet can't mount the volume with `-o context=c1,c2`, because the volume is already mounted.
+   It needs to wait for Pod A to get deleted + its volume unmounted.
+
+### Make SELinuxMount opt-in and not opt-out
+
+I.e. the old behavior is still the default.
+
+* Safer, there is no behavior change. Users that want to benefit from the new feature must opt-in.
+* Mounting with SELinux is safe for RWOP volumes, and therefore we would like to have it enabled by default.
+  The resulting `SELinuxChangePolicy` values would be:
+  * `MountReadWriteOncePod` - mount RWOP volumes with SELinux mount option, recursive relabeling for the rest. This is the default.
+  * `Recursive` - explicit opt-out.
+  * `MountOption` - mount all volumes with SELinux mount option, needs explicit opt-in.
+  * We consider such API harder to understand.
+
+### Allow opt-in (or opt-out) globally via kubelet flags
+
+Instead of `SELinuxChangePolicy` in every PodSpec, we considered having a global flag to enable or disable the feature for all Pods in a cluster.
+For example by a new kubelet flag `--enable-selinux-mount`.
+
+This is a good "plan B", if the API changes proposed here are not acceptable.
+
+
+# Appendix 1: promQL queries
+```yaml
+apiVersion: monitoring.coreos.com/v1
+kind: PrometheusRule
+metadata:
+  name: selinuxmount-test
+spec:
+  groups:
+    - name: selinuxmount.rules
+      rules:
+      # Add unique_volume_id label to kube_persistentvolume_info that identifies a volume in the storage backend.
+      # PVs A and B represent the same volume in the storage backend if and only if they have the same unique_volume_id.
+      # The string is not very human friendly.
+      - record: kube_persistentvolume_info:unique_volume_id
+        expr: |
+          label_join(kube_persistentvolume_info, "unique_volume_id", "#", "gce_persistent_disk_name", "ebs_volume_id", "azure_disk_name", "fc_wwids", "fc_lun", "fc_target_wwns", "iscsi_target_portal", "iscsi_iqn", "iscsi_lun", "nfs_server", "nfs_path", "csi_driver", "csi_volume_handle", "local_path", "host_path")
```

Comment:
it seems difficult to maintain? wouldn't it be hard to know all the label names in advance?

Reply:
+1, this adds value only if the querying interface has a dropdown, in addition to the string itself being hard to read and write. The cardinality, which seems high and unbounded, stays the same if this is split into labels, which makes things convenient from the UX perspective (allowing the user to filter by `fc_wwids`, etc.).

Reply:
I'd prefer having this as a metric in KSM where the value would be a hash instead, to counter both UX and cardinality issues.

Reply:
I moved the kube-state-metrics to "Alternatives" (a new controller in KCM with metrics + events is the new proposal).